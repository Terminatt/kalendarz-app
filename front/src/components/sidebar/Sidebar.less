@import '~styles/variables.less';
@import '~styles/breakpoints.less';

.sidebar {
    position: relative;
    height: 100%;
    z-index: 10;
    &-wrapper {
        position: absolute;
        background: @primary-bg;
        width: @sidebar-width;
        height: 100%;

        &-close { 
            position: absolute;
            top: 20px;
            right: 15px;
        }
    
        &-header {
            padding: @padding-big @padding-medium 0 @padding-medium;
    
            &-content {
                margin-bottom: @margin-tiny;
    
                &-link {
                    padding: 0 @padding-tiny @padding-tiny @padding-tiny;
                    color: @text-color;
                }
        
                &-link:focus {
                    outline: 2px solid @border-color-base;
                    width: max-content;
                    height: max-content;
                }
    
                &-link:hover {
                    color: @text-color;
                    text-decoration: underline;
                }
            }
        }
    
        &-top {
            padding: 0 @padding-medium;
        }
    
        &-bottom {
            padding: 0 @padding-medium;
        }
    }

}

.site-mask {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    width: 100vw;
    height: 100vh;
    background-color: @black;
    opacity: @mask-opacity;
}

@media screen and (min-width: @w-between) {
    .sidebar {
        &-wrapper {
            position: relative;
            border-radius: @border-radius-regular;
            box-shadow: @box-shadow-mild;

            &-close {
                display: none;
            }
        }
    }
    
    .site-mask {
        display: none;
    }
}