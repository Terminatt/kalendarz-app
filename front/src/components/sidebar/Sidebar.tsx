import React from 'react';
import { Divider } from 'antd';
import { GenericReactContent } from '@generics/generics';
import { Link } from 'react-router-dom';

import './Sidebar.less';

export interface SidebarProps {
    top: GenericReactContent;
    bottom?: GenericReactContent;
    headerText?: string | GenericReactContent
}

const Sidebar: React.FC<SidebarProps> = (props) => {
    const { top, bottom, headerText } = props;

    return (
        <div className="sidebar">
            <header className="sidebar-header">
                <h1 className="sidebar-header-content">
                    <Link to="/" tabIndex={0} className="sidebar-header-content-link">
                        {headerText}
                    </Link>
                </h1>
            </header>
            <div className="sidebar-top">
                {top}
            </div>
            {!!bottom && <Divider />}
            <div className="sidebar-bottom">
                {bottom}
            </div>
        </div>
    );
};

export default Sidebar;
