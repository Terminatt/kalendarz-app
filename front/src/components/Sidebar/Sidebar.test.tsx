import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import Sidebar from './Sidebar';

describe('Sidebar Component', () => {
    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <MemoryRouter>
                    <Sidebar headerText="This is a text" visible top={<div>Top Test</div>} bottom={<div>Bottom Test</div>} />
                </MemoryRouter>
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('renders the content properly', async () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';
        const headerText = "Header text";

        const element = render(
            <MemoryRouter>
                <Sidebar headerText={headerText} visible top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );
        expect(await element.queryByText(topText)).not.toBeNull();
        expect(await element.queryByText(bottomText)).not.toBeNull();
        expect(await element.queryByText(headerText)).not.toBeNull();

    });

    it('does not display when small screen and not visible', async () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';

        const element = render(
            <MemoryRouter>
                <Sidebar isSmallScreen visible={false} top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );

        expect(await element.queryByTestId('sidebar')).toBeNull();
    });

    it('is displayed when isSmallScreen', async () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';

        const element = render(
            <MemoryRouter>
                <Sidebar isSmallScreen visible={true} top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );

        expect(await element.queryByTestId('sidebar')).not.toBeNull()
    });

    it('does not have display none when visible', () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';

        const element = render(
            <MemoryRouter>
                <Sidebar visible top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );

        expect(element.getByTestId('sidebar')).not.toHaveStyle('display: none;');
    });

    it('renders close button when small screen', async () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';

        const element = render(
            <MemoryRouter>
                <Sidebar isSmallScreen visible top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );

        expect(await element.queryByTestId('close-btn')).not.toBeNull;
    });

    it('triggers on close', async () => {
        const topText = 'Top Test';
        const bottomText = 'Bottom Test';
        const onClose = jest.fn()

        const element = render(
            <MemoryRouter>
                <Sidebar onClose={onClose} isSmallScreen visible top={<div>{topText}</div>} bottom={<div>{bottomText}</div>}  />
            </MemoryRouter>
        );
        
        fireEvent.click(await element.findByTestId('close-btn'))
        expect(onClose).toHaveBeenCalledTimes(1);
    });
});
