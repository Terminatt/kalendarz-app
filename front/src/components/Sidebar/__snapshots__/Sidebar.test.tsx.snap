// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component matches the snapshot 1`] = `
<aside
  className="sidebar"
  data-testid="sidebar"
  style={
    Object {
      "display": "",
    }
  }
>
  <div
    className="sidebar-wrapper"
  >
    <header
      className="sidebar-wrapper-header"
    >
      <h1
        className="sidebar-wrapper-header-content"
      >
        <a
          className="sidebar-wrapper-header-content-link"
          href="/"
          onClick={[Function]}
          tabIndex={0}
        >
          This is a text
        </a>
      </h1>
    </header>
    <div
      className="sidebar-wrapper-top"
    >
      <div>
        Top Test
      </div>
    </div>
    <div
      className="ant-divider ant-divider-horizontal"
      role="separator"
    />
    <div
      className="sidebar-wrapper-bottom"
    >
      <div>
        Bottom Test
      </div>
    </div>
  </div>
  <div
    className="mask sidebar-mask"
  />
</aside>
`;
